pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/SegmentDeletionManager.java

```diff
@@ -71,108 +71,124 @@ public void stop() {
+        String segmentPropertyStorePath = ZKMetadataProvider.constructPropertyStorePathForSegment(tableName, segmentId);
+        propStorePathList.add(segmentPropertyStorePath);
       }
 
-      if (!deletionFromPropertyStoreSuccessful) {
-        long effectiveDeletionDelay = Math.min(deletionDelay * 2, MAX_DELETION_DELAY_SECONDS);
-        LOGGER.warn("Failed to delete segment {} from property store, will retry in {} seconds.", segmentId,
-            effectiveDeletionDelay);
-        deleteSegmentWithDelay(tableName, segmentId, effectiveDeletionDelay);
-        return;
+      boolean[] status = _propertyStore.remove(propStorePathList, AccessOption.PERSISTENT);
+      int nPropstoreDeleteFailed = 0;
+      for (int i = 0; i < status.length; i++) {
+        if (!status[i]) {
+          // remove API can fail because the prop store entry did not exist, so check first.
+          if (_propertyStore.exists(propStorePathList.get(i), AccessOption.PERSISTENT)) {
+            nPropstoreDeleteFailed++;
+            LOGGER.info("Could not delete {} from propertystore", propStorePathList.get(i));
+            segmentsToRetryLater.add(segmentsToDelete.get(i));
+          }
+        }
       }
 
-      final String rawTableName = TableNameBuilder.extractRawTableName(tableName);
-      if (_localDiskDir != null) {
-        File fileToMove = new File(new File(_localDiskDir, rawTableName), segmentId);
-        if (fileToMove.exists()) {
-          File targetDir = new File(new File(_localDiskDir, DELETED_SEGMENTS), rawTableName);
-          try {
-            // Overwrites the file if it already exists in the target directory.
-            FileUtils.copyFileToDirectory(fileToMove, targetDir, true);
-            LOGGER.info("Moved segment {} from {} to {}", segmentId, fileToMove.getAbsolutePath(),
-                targetDir.getAbsolutePath());
-            if (!fileToMove.delete()) {
-              LOGGER.warn("Could not delete file", segmentId, fileToMove.getAbsolutePath());
-            }
-          } catch (IOException e) {
-            LOGGER.warn("Could not move segment {} from {} to {}", segmentId, fileToMove.getAbsolutePath(),
-                targetDir.getAbsolutePath(), e);
-          }
-        } else {
-          CommonConstants.Helix.TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
-          switch (tableType) {
-            case OFFLINE:
-              LOGGER.warn("Not found local segment file for segment {}" + fileToMove.getAbsolutePath());
-              break;
-            case REALTIME:
-              if (SegmentName.isLowLevelConsumerSegmentName(segmentId)) {
-                LOGGER.warn("Not found local segment file for segment {}" + fileToMove.getAbsolutePath());
-              }
-              break;
-            default:
-              LOGGER.warn("Unsupported table type {} when deleting segment {}", tableType, segmentId);
+      for (String segmentId : segmentsToDelete) {
```

Comment:
We discussed, and concluded that we will have a race condition no matter what, since we do not use a (zk-based) lock to exclude addition of a segment and its deletion (and fetch, of course). However, this is not likely to happen because our operational guide to replace a segment is to simply push the new segment (and not drop the old one and then push the new one).

That said, a corner case does exist if someone chooses to drop old segment and then push the new one by same name. For example, it may happen that the delete is postponed several minutes due to the segment not disappearing form externalview (known to have happened in large clusters), EV gets corrected, and the user believes the segment is gone, and then uploads the new segement.